File: src/types/memory.ts
/**
 * @module types/memory
 * @description TypeScript types for the Memory Management feature
 *
 * PURPOSE:
 * - Define memory source, learning, and health types
 * - Mirror Rust models in src-tauri/src/models/memory.rs
 *
 * EXPORTS:
 * - MemorySource - A memory file (CLAUDE.md, rules, skills, etc.)
 * - Learning - An extracted learning with metadata
 * - MemoryHealth - Overall memory health metrics
 * - ClaudeMdAnalysis - CLAUDE.md quality analysis
 * - AnalysisSuggestion - Improvement suggestion
 * - LineRemovalSuggestion - Line removal recommendation
 * - LineMoveTarget - Line move recommendation
 * - LearningCategory, LearningTopic, LearningStatus, ConfidenceLevel, HealthRating types
 *
 * PATTERNS:
 * - Import from '@/types/memory' for memory-specific types
 * - Or import from '@/types' for convenience re-exports
 *
 * CLAUDE NOTES:
 * - Keep in sync with Rust models in src-tauri/src/models/memory.rs
 * - MemorySource.sourceType values: "claude-md", "rules", "skills", "local-md", etc.
 * - HealthRating is derived from claudeMdScore and overall metrics
 */

export type LearningCategory = 'Preference' | 'Solution' | 'Pattern' | 'Gotcha';
export type LearningTopic = 'debugging' | 'patterns' | 'tools' | 'project' | 'workflow';
export type LearningStatus = 'active' | 'verified' | 'deprecated' | 'archived';
export type ConfidenceLevel = 'high' | 'medium' | 'low';
export type HealthRating = 'excellent' | 'good' | 'needs-attention' | 'poor';

export interface MemorySource {
  path: string;
  sourceType: string;
  name: string;
  lineCount: number;
  sizeBytes: number;
  lastModified: string;
  description: string;
}

export interface Learning {
  id: string;
  sessionId: string;
  category: LearningCategory;
  content: string;
  topic: LearningTopic | null;
  confidence: ConfidenceLevel;
  status: LearningStatus;
  sourceFile: string;
  createdAt: string;
  updatedAt: string;
}

export interface MemoryHealth {
  totalSources: number;
  totalLines: number;
  totalLearnings: number;
  activeLearnings: number;
  claudeMdLines: number;
  claudeMdScore: number;
  rulesFileCount: number;
  skillsCount: number;
  estimatedTokenUsage: number;
  healthRating: HealthRating;
}

export interface ClaudeMdAnalysis {
  totalLines: number;
  estimatedTokens: number;
  score: number;
  sections: string[];
  suggestions: AnalysisSuggestion[];
  linesToRemove: LineRemovalSuggestion[];
  linesToMove: LineMoveTarget[];
}

export interface AnalysisSuggestion {
  suggestionType: string;
  message: string;
  lineRange: [number, number] | null;
  target: string | null;
}

export interface LineRemovalSuggestion {
  lineNumber: number;
  content: string;
  reason: string;
}

export interface LineMoveTarget {
  lineRange: [number, number];
  contentPreview: string;
  targetFile: string;
  reason: string;
}

File: src-tauri/src/models/memory.rs
//! @module models/memory
//! @description Data models for the Memory Management feature
//!
//! PURPOSE:
//! - Define memory source, learning, and memory health data structures
//! - Provide serialization for Tauri IPC
//!
//! DEPENDENCIES:
//! - serde - Serialization/deserialization for Tauri IPC
//!
//! EXPORTS:
//! - MemorySource - Represents a memory file (CLAUDE.md, rules, auto-memory, etc.)
//! - Learning - An extracted learning with category, topic, confidence, status
//! - MemoryHealth - Overall memory health metrics
//! - ClaudeMdAnalysis - Analysis results for CLAUDE.md quality
//! - AnalysisSuggestion - Individual suggestion for CLAUDE.md improvement
//! - LineRemovalSuggestion - Suggestion to remove a specific line
//! - LineMoveTarget - Suggestion to move lines to another file
//!
//! PATTERNS:
//! - All models derive Serialize, Deserialize for Tauri IPC
//! - Uses #[serde(rename_all = "camelCase")] for TS interop
//! - Keep in sync with TypeScript types in src/types/memory.ts
//!
//! CLAUDE NOTES:
//! - MemorySource.source_type values: "claude-md", "rules", "auto-memory", "local", "skills"
//! - MemorySource.scope values: "project", "global"
//! - Learning.category values: "Preference", "Solution", "Pattern", "Gotcha"
//! - Learning.confidence values: "high", "medium", "low"
//! - Learning.status values: "active", "verified", "deprecated", "archived"
//! - MemoryHealth.health_rating values: "excellent", "good", "needs-attention", "poor"

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemorySource {
    pub path: String,
    pub source_type: String,
    pub name: String,
    pub scope: String,
    pub line_count: u32,
    pub size_bytes: u64,
    pub last_modified: String,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Learning {
    pub id: String,
    pub session_id: String,
    pub category: String,
    pub content: String,
    pub topic: Option<String>,
    pub confidence: String,
    pub status: String,
    pub source_file: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryHealth {
    pub total_sources: u32,
    pub total_lines: u32,
    pub total_learnings: u32,
    pub active_learnings: u32,
    pub claude_md_lines: u32,
    pub claude_md_score: u32,
    pub rules_file_count: u32,
    pub skills_count: u32,
    pub estimated_token_usage: u32,
    pub health_rating: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClaudeMdAnalysis {
    pub total_lines: u32,
    pub estimated_tokens: u32,
    pub score: u32,
    pub sections: Vec<String>,
    pub suggestions: Vec<AnalysisSuggestion>,
    pub lines_to_remove: Vec<LineRemovalSuggestion>,
    pub lines_to_move: Vec<LineMoveTarget>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisSuggestion {
    pub suggestion_type: String,
    pub message: String,
    pub line_range: Option<(u32, u32)>,
    pub target: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LineRemovalSuggestion {
    pub line_number: u32,
    pub content: String,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LineMoveTarget {
    pub line_range: (u32, u32),
    pub content_preview: String,
    pub target_file: String,
    pub reason: String,
}
